src/Controller/Controller.php

```diff
@@ -492,12 +492,17 @@ public function setResponse(Response $response)
      * Dispatches the controller action. Checks that the action
      * exists and isn't private.
      *
-     * @return \Psr\Http\Message\ResponseInterface The resulting response.
+     * @return \Psr\Http\Message\ResponseInterface
      * @throws \Cake\Controller\Exception\MissingActionException If controller action is not found.
      * @throws \UnexpectedValueException If return value of action method is not null or ResponseInterface instance.
      */
-    public function invokeAction(): ?ResponseInterface
+    public function invokeAction(): ResponseInterface
```

Comment:
This is a good simplification of the interface. I am left wondering if it is too opaque though for future plans. We have talked about adding a DI container in 4.x and injecting parameters into controller actions. This interface would require us to give the controller a reference to the container which isn't a pattern I like.

Perhaps instead of invokeAction() controllers should have a method that returns a Closure with the same signature as the action which could allow us to use a container to inject parameters more easily.

There are likely other/better solutions but I would like to avoid giving controllers a reference to the DI container.